unity-renderer/Assets/Scripts/MainScripts/DCL/WorldRuntime/KernelCommunication/RPC/Services/EmotesService/EmotesRendererServiceImpl.cs

```diff
@@ -1,22 +1,178 @@
+
+            port.OnClose += OnPortClosed;
+        }
+
+        private void OnPortClosed()
+        {
+            foreach (var scenes in worldState.GetLoadedScenes())
+            {
+                OnSceneRemoved(scenes.Value);
+            }
         }
 
         public async UniTask<TriggerSelfUserExpressionResponse> TriggerSelfUserExpression(TriggerSelfUserExpressionRequest request, RPCContext context, CancellationToken ct)
         {
             await UniTask.SwitchToMainThread(ct);
-            UserProfile.GetOwnUserProfile().SetAvatarExpression(request.Id, UserProfile.EmoteSource.Command);
+            userProfile.SetAvatarExpression(request.Id, UserProfile.EmoteSource.Command);
             return default;
         }
+
+        public async UniTask<SuccessResponse> TriggerSceneExpression(TriggerSceneExpressionRequest request, RPCContext context, CancellationToken ct)
+        {
+            // try get loaded scene
+            if (!worldState.TryGetScene(request.SceneNumber, out IParcelScene scene))
+                return FAILURE_RESPONSE;
+
+            // generates an emote id
+            if (!SceneEmoteHelper.TryGenerateEmoteId(scene, request.Path, request.Loop, out string emoteId))
+                return FAILURE_RESPONSE;
+
+            avatarData ??= ownPlayer.Get()?.avatar;
+
+            if (avatarData == null)
+                return FAILURE_RESPONSE;
+
+            string userBodyShape = userProfile.avatar.bodyShape;
+
+            // get hashset for scene emotes that are currently loading or create one if none
+            if (!pendingEmotesByScene.TryGetValue(scene, out HashSet<(string bodyshapeId, string emoteId)> scenePendingEmotes))
+            {
+                scenePendingEmotes = new HashSet<(string bodyshapeId, string emoteId)>();
+                pendingEmotesByScene.Add(scene, scenePendingEmotes);
+            }
+
+            // get hashset for scene emotes that are already equipped
+            if (!equippedEmotesByScene.TryGetValue(scene, out HashSet<(string bodyshapeId, string emoteId)> sceneEquippedEmotes))
+            {
+                sceneEquippedEmotes = new HashSet<(string bodyshapeId, string emoteId)>();
+                equippedEmotesByScene.Add(scene, sceneEquippedEmotes);
+            }
+
+            sceneController.OnSceneRemoved -= OnSceneRemoved;
+            sceneController.OnSceneRemoved += OnSceneRemoved;
+
+            try
+            {
+                await UniTask.SwitchToMainThread(ct);
+
+                // request emote to load using DataStore::emotes which will make `EmoteAnimationsTracker` plugin to load the emote
+                await SceneEmoteHelper.RequestLoadSceneEmote(
+                    userBodyShape,
+                    emoteId,
+                    alreadyLoadedEmotes,
+                    emotesInUse,
+                    scenePendingEmotes,
+                    sceneEquippedEmotes,
+                    () => !worldState.ContainsScene(request.SceneNumber),
```

Comment:
I truly believe there should be a reliable `CancellationTokenSource` corresponding to the scene

Reply:
done https://github.com/decentraland/unity-renderer/pull/5501/commits/04a3ef16362e9dda561cbc2cda137ea51def5979